[DWARF] Use ULEB128 and not just one byte for directory indices

drodriguez opened the pull request:
According to the standard `DW_LNCT_directory_index` can be `data1`, `data2`, or `udata` (see 6.2.4.1). The code was using `data1`, but this limits the number of directories to 256, even if the variable holding the directory index is a `uint64_t`. `dsymutil` was hitting an assertion when trying to write directory indices higher than 255.

Modify the classic and the parallel DWARF linkers to use `udata` and encode the directory indices as ULEB128 and provide a test that has more than 256 directories to check the changes are working as expected.

For people that were using `dsymutil` with CUs that had between 128-256 directories, this will mean that for those indices 2 bytes will be used now, instead of just one.

drodriguez committed:
[DWARF] Use ULEB128 and not just one byte for directory indices

According to the standard `DW_LNCT_directory_index` can be `data1`, `data2`,
or `udata` (see 6.2.4.1). The code was using `data1`, but this limits the
number of directories to 256, even if the variable holding the directory
index is a `uint64_t`. `dsymutil` was hitting an assertion when trying
to write directory indices higher than 255.

Modify the classic and the parallel DWARF linkers to use `udata` and
encode the directory indices as ULEB128 and provide a test that has more
than 256 directories to check the changes are working as expected.

For people that were using `dsymutil` with CUs that had between 128-256
directories, this will mean that for those indices 2 bytes will be used
now, instead of just one.

ellishg commented on llvm/test/tools/dsymutil/X86/dwarf5-many-include-directories.ll:
```diff
@@ -0,0 +1,2631 @@
+; RUN: rm -rf %t && mkdir -p %t
+; RUN: %llc_dwarf -o %t/all.o -filetype=obj %s
+; Dummy debug map to get the results. Bare object doesn't seem to work
+; RUN: echo "---" > %t/debug.map
+; RUN: echo "triple: 'x86_64-apple-darwin'" >> %t/debug.map
+; RUN: echo "objects:" >> %t/debug.map
+; RUN: echo "  - filename: %t/all.o" >> %t/debug.map
+; RUN: echo "    symbols:" >> %t/debug.map
+; RUN: echo "      - { sym: _all, objAddr: 0x0, binAddr: 0x0, size: 0x0 }" >> %t/debug.map
+; RUN: echo "..." >> %t/debug.map
```
`split-file` would allow you to write `debug.map` inline in the file without using `echo`. I've used `sed` to write the temporary file path into these files before.

https://llvm.org/docs/TestingGuide.html#extra-files

ellishg commented on llvm/test/tools/dsymutil/X86/dwarf5-many-include-directories.ll:
```diff
@@ -0,0 +1,2631 @@
+; CHECK: dir_index: 257
+
+; ---
+; Generated doing (fish shell):
+; - for cnt in (seq 0 256); mkdir -p d$cnt ; printf "void func$cnd() {}\n#define FUNC$cnt func$cnt()\n" >> d$cnt/f$cnt.c ; end
+; - for cnt in (seq 0 256); printf "#include \"f$cnt.c\"" >> all.c ; end
+; - printf "void all() {\n" >> all.c
+; - for cnt in (seq 0 256); printf "FUNC$cnt;\n" >> all.c ; end
+; - printf "}\n" >> all.c
+; - clang -target x86_64-apple-macos -S -emit-llvm -gdwarf-5 -o all.ll all.c (for cnt in (seq 0 256); echo "-Id$cnt"; end)
+; - Edit all.ll manually and change all DIFile so the directory in filename is
+;   moved into the directory field.
```
It would be nice to use Python to generate this `.ll` file so we don't have to checkin such a large file. Here is an example where I've done this.

https://github.com/ellishg/llvm-project/blob/e40915b7407eda4b370658da5c9606e310b55d19/lld/test/MachO/bp-section-orderer-stress.s#L5-L17

drodriguez committed:
Remake the test in Python to avoid large test.

- Use `.test` as the extension. Remove `.ll` from test extensions.
- The output generated by Python is the same, except for the checksum
  fields, which are constant, but does not seem to matter.

drodriguez committed:
Make sed regex compatible with Windows.

drodriguez committed:
Force use of Mach-O format for dsymutil, even on Windows

kyulee-com reviewed (approved):
The code looks good to me!
Thanks for  creating a tricky test that requires large files from a script.
While I'm approving this, I would also like to give @JDevlieghere,  a chance to review.

JDevlieghere reviewed (approved):
Given that we know the number of indices, we could always emit the most efficient encoding. That said, we're emitting this only once per line table, so I'm not particularly worried about the additional byte and simplifies the implementation. LGTM.

drodriguez commented:
> Given that we know the number of indices, we could always emit the most efficient encoding. That said, we're emitting this only once per line table, so I'm not particularly worried about the additional byte and simplifies the implementation. LGTM.

I was striving for implementation simplicity. I started to think about a better implementation with different formats, but I did not see clearly how it would be better.

```
| Range         | data1    | data2     | udata     |
| 0-127         | 1 byte * | 2 bytes   | 1 byte *  |
| 128-255       | 1 byte * | 2 bytes   | 2 bytes   |
| 255-16383     | NA       | 2 bytes * | 2 bytes * |
| 16384-65535   | NA       | 2 bytes * | 3 bytes   |
| 65536-2097151 | NA       | NA        | 3 bytes * | 
```

One thing to notice is that one can use the range to choose a format, and that works for getting the best format for 0-16383, but for 16384-65536, one needs to calculate the distribution of indices to see if the udata encoding or the data2 encoding will be smaller.
